### PR TITLE
add bbox_overlaps for fp16 to save memory and keep speed

### PR DESCRIPTION
**Pull request by li-phone**
In GPU and CPU environments with limited hardware resources, bbox_overlaps often has CUDA memory allocation errors (there are a large number of anchor boxes when calculating IOU). To do this, we added bbox_overlaps on FP16 to save memory and keep speed at the same time. 
TO DO: 
all IOU calculation on fp16 including NMS.

### CODE REVIEW

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py [hunk 1]
```diff
@@ -3,10 +3,28 @@
 from .builder import IOU_CALCULATORS
 
 
+def cast_tensor_type(x, scale=1., dtype=None):
+    if dtype == 'fp16':
+        x = (x / scale).half()
+    return x
+
+
+def __clamp__(x, min=None, max=None):
```
Comment (ZwwWayne) [inline mmdet/core/bbox/iou_calculators/iou2d_calculator.py:12]: suggest to name it as `fp16_clamp`

**Comment by ZwwWayne** [comment]
Hi @li-phone ,
Thanks for the PR. This feature is quite interesting and valuable, so we plan to merge it in the next release (end of April). Could you sign the CLA and fix the lint/CI issue? Furthermore, we suggest adding unit tests for the FP16 feature to ensure correctness, thus we do not need to train models to do that.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py (hunk 1 above)
Comment (li-phone) [reply mmdet/core/bbox/iou_calculators/iou2d_calculator.py]: yeah, that sounds very good

**Comment by ZwwWayne** [comment]
Hi @li-phone ,
Thanks for the quick response. Do you have some statistics about how much the memory and the time cost are reduced? 
If you have that, could you post them in your PR message and add some comments in the code you modified, thus future developers could recognize the value of the modification. We plan to merge this PR for the next release (end of this month)

**Comment by li-phone** [comment]
> Hi @li-phone ,
> Thanks for the quick response. Do you have some statistics about how much the memory and the time cost are reduced?
> If you have that, could you post them in your PR message and add some comments in the code you modified, thus future developers could recognize the value of the modification. We plan to merge this PR for the next release (end of this month)

Yes, we proved it as follows: 
**Prove:**
        Assume bboxes1 is M x 4, bboxes2 is N x 4, when mode is 'iou', there are some **new generated variable** when calculating IOU using bbox_overlaps function:
        1) is_aligned is False
            area1: M x 1
            area2: N x 1
            lt: M x N x 2
            rb: M x N x 2
            wh: M x N x 2
            overlap: M x N x 1
            union: M x N x 1
            ious: M x N x 1

            Total memory:
                S = (9 x N x M + N + M) * 4 Byte

            When using FP16, we can reduce:
                R = (9 x N x M + N + M) * 4 / 2 Byte,
                R large than (N + M) * 4 * 2 **is always true** when N and M >= 1.
                Obviously, N + M <= N * M < 3 * N * M, when N >=2 and M >=2,
                                  N + 1 < 3 * N, when N or M is 1.


            Given M = 40 (ground truth), N = 400000 (three anchor boxes
            in per grid, FPN, R-CNNs),
                R = 275 MB (one times)

            A special case (**dense detection**), M = 512 (ground truth),
                R = 3516 MB = 3.43 GB

            When the batch size is B, reduce:
                B x R

            Therefore, **CUDA memory runs out frequently**.

            Experiments on GeForce RTX 2080Ti (11019 MiB):

            |   dtype   |   M   |   N   |   Use    |   Real   |   Ideal   |
            |:----:|:----:|:----:|:----:|:----:|:----:|
            |   FP32   |   512 | 400000 | 8020 MiB |   --   |   --   |
            |   FP16   |   512 | 400000 |   4504 MiB | **3516 MiB** | **3516 MiB** |
            |   FP32   |   40 | 400000 |   1540 MiB |   --   |   --   |
            |   FP16   |   40 | 400000 |   1264 MiB |   **276MiB**   | **275 MiB** |

        2) is_aligned is True
            area1: N x 1
            area2: N x 1
            lt: N x 2
            rb: N x 2
            wh: N x 2
            overlap: N x 1
            union: N x 1
            ious: N x 1

            Total memory:
                S = 11 x N * 4 Byte

            When using FP16, we can reduce:
                R = 11 x N * 4 / 2 Byte

        So do the 'giou' (**large than 'iou'**).

        Time-wise, **FP16 is generally faster than FP32.**

        When gpu_assign_thr is not -1, **it takes more time on cpu
        but not reduce memory.**
        Therefore, **we can reduce half the memory and keep the speed.**

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py [hunk 2]
```diff
@@ -32,17 +51,99 @@ def __call__(self, bboxes1, bboxes2, mode='iou', is_aligned=False):
...
-        repr_str = self.__class__.__name__ + '()'
+        repr_str = self.__class__.__name__ + f'(' \
+            f'scale={self.scale}, dtype={self.dtype})'
         return repr_str
 
 
 def bbox_overlaps(bboxes1, bboxes2, mode='iou', is_aligned=False, eps=1e-6):
     """Calculate overlap between two set of bboxes.
 
+    FP16 Contributed by: https://github.com/li-phone
```
Comment (ZwwWayne) [inline mmdet/core/bbox/iou_calculators/iou2d_calculator.py:77]: 1. May refer to the PR ID.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py [hunk 3]
```diff
@@ -32,17 +51,99 @@ def __call__(self, bboxes1, bboxes2, mode='iou', is_aligned=False):
...
+    Prove:
```
Comment (ZwwWayne) [inline mmdet/core/bbox/iou_calculators/iou2d_calculator.py:78]: `Prove` is not a standard section in the docstring, we may use `Note`

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py [hunk 4]
```diff
@@ -152,7 +255,8 @@ def bbox_overlaps(bboxes1, bboxes2, mode='iou', is_aligned=False, eps=1e-6):
     if mode in ['iou', 'iof']:
         return ious
     # calculate gious
-    enclose_wh = (enclosed_rb - enclosed_lt).clamp(min=0)
+    # enclose_wh = (enclosed_rb - enclosed_lt).clamp(min=0)
```
Comment (ZwwWayne) [inline mmdet/core/bbox/iou_calculators/iou2d_calculator.py:258]: delete the commented unnecessary code.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py [hunk 5]
```diff
@@ -133,7 +235,8 @@ def bbox_overlaps(bboxes1, bboxes2, mode='iou', is_aligned=False, eps=1e-6):
         rb = torch.min(bboxes1[..., :, None, 2:],
                        bboxes2[..., None, :, 2:])  # [B, rows, cols, 2]
 
-        wh = (rb - lt).clamp(min=0)  # [B, rows, cols, 2]
+        # wh = (rb - lt).clamp(min=0)  # [B, rows, cols, 2]
```
Comment (ZwwWayne) [inline mmdet/core/bbox/iou_calculators/iou2d_calculator.py:238]: delete the commented unnecessary code.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py [hunk 6]
```diff
@@ -32,17 +51,99 @@ def __call__(self, bboxes1, bboxes2, mode='iou', is_aligned=False):
             bboxes2 = bboxes2[..., :4]
         if bboxes1.size(-1) == 5:
             bboxes1 = bboxes1[..., :4]
+
+        if self.dtype == 'fp16':
+            # change tensor type to save cpu and cuda memory and keep speed
+            bboxes1 = cast_tensor_type(bboxes1, self.scale, self.dtype)
```
Comment (ZwwWayne) [inline mmdet/core/bbox/iou_calculators/iou2d_calculator.py:57]: Can we move the logic of lines 55-63 into `bbox_overlaps`? This makes `bbox_overlaps` a more general function that enables FP16 and simplifies the logic in BboxOverlaps2D.__call__.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py (hunk 2 above)
Comment (li-phone) [reply mmdet/core/bbox/iou_calculators/iou2d_calculator.py:77]: Okay, No problem.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py (hunk 6 above)
Comment (li-phone) [reply mmdet/core/bbox/iou_calculators/iou2d_calculator.py:57]: enen, thank you for your advice.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py (hunk 5 above)
Comment (li-phone) [reply mmdet/core/bbox/iou_calculators/iou2d_calculator.py:238]: enen, thank you for your advice.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py (hunk 4 above)
Comment (li-phone) [reply mmdet/core/bbox/iou_calculators/iou2d_calculator.py:258]: enen, thank you for your advice.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py (hunk 3 above)
Comment (li-phone) [reply mmdet/core/bbox/iou_calculators/iou2d_calculator.py:78]: enen, thank you for your advice.

File: mmdet/core/bbox/iou_calculators/iou2d_calculator.py (hunk 6 above)
Comment (li-phone) [reply mmdet/core/bbox/iou_calculators/iou2d_calculator.py:57]: yeah, of course you can.

Comment (ZwwWayne) [reply mmdet/core/bbox/iou_calculators/iou2d_calculator.py:57]: OK  agree

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**